unknown_type/[M]230.cpp

```diff
@@ -0,0 +1,33 @@
+// Problem: 230. Kth Smallest Element in a BST
+// Link: https://leetcode.com/problems/kth-smallest-element-in-a-bst/
+
+/* Nhận xét:
+- Với một BST thì ta duyệt theo inorder traversal sẽ được một dãy tăng dần.
+- Mỗi khi duyệt đến một đỉnh, ta giảm k đi 1. Khi nào giảm tới 0 thì trả về kết quả.
+- Time complexity: O(max(log(n), k)), do phải mất log(n) bước để tìm phần tử nhỏ nhất, sau đó tìm k đỉnh tiếp theo. Trường hợp suy biến là O(n) khi BST bị suy biến.
+- Space complexity: O(1).
+*/
+
+class Solution {
+public:
+    void traverse(TreeNode *u, int &k, int &res) {
+        if (u) {
+            if (k > 0) {
```

Comment:
Check if (k > 0) này ko cần thiết. Cả ở dòng 23 cũng ko cần thiết.

Reply:
Anh kiểm tra thì nếu bỏ đi thì chương trình sẽ phải duyệt gần như toàn bộ cây, kể cả khi đã tìm ra kết quả.
Như vậy độ phức tạp sẽ trở thành O(n) là lớn hơn trước.
Vậy có ảnh hưởng gì ko?

Reply:
Nếu vậy chú có thể đặt trên đầu hàm : if (!u || k == 0) return; Nó sẽ tránh được trường hợp duyệt cả cây.

Reply:
Ok, got it! Fixed!

Reply:
Ok code vừa ngon vừa thon vừa gọn rồi đấy :)) Tổng quát cho các bài recursion, chú nên check trường hợp tầm thường ở đầu hàm rồi return ngay lập tức. Đó sẽ là cách code tốt.

Reply:
Thanks bro. Got it!